primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,399 @@
+pub struct StorageLock<'a, L>
+where
+	L: Sized + Lockable,
+{
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	deadline: L,
+}
+
+impl<'a, L> StorageLock<'a, L>
+where
+	L: Lockable,
+{
+	/// Create a new storage lock with [default expiry duration](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+	pub fn new<'k>(key: &'k [u8]) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: L::deadline(),
+		}
+	}
+
+	#[inline]
+	fn try_lock_inner(&mut self, new_deadline: L) -> Result<(), Option<L>> {
+		let now = L::current();
+		let res = self
+			.value_ref
+			.mutate(|s: Option<Option<L>>| -> Result<L, Option<L>> {
+				match s {
+					// no lock set, we can safely acquire it
+					None => Ok(new_deadline),
+					// lock is set, but it's old. We can re-acquire it.
+					Some(Some(deadline)) if now.expired(&deadline) => Ok(new_deadline),
+					// lock is present and is still active
+					Some(Some(deadline)) => Err(Some(deadline)),
+					_ => Err(None),
+				}
+			});
+		match res {
+			Ok(Ok(_)) => Ok(()),
+			Ok(Err(_deadline)) => Err(None),
+			Err(e) => Err(e),
+		}
+	}
+
+	/// Attempt to lock the storage entry.
+	///
+	/// Returns a lock guard on success, otherwise an error containing `None` in
+	/// case the mutex was already unlocked before, or if the lock is still held
+	/// by another process `Err(Some(expiration_timestamp))`.
+	pub fn try_lock<'b>(&'b mut self) -> Result<StorageLockGuard<'a, 'b, L>, Option<L>>
+	where
+		'a: 'b,
+	{
+		match self.try_lock_inner(self.deadline) {
+			Ok(_) => Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+			Err(e) => Err(e),
+		}
+	}
+
+	/// Try grabbing the lock until its expiry is reached.
+	///
+	/// Returns an error if the lock expired before it could be caught.
+	pub fn spin_lock<'b, 'c>(&'b mut self) -> StorageLockGuard<'a, 'b, L>
+	where
+		'a: 'b,
+	{
+		loop {
+			// blind attempt on locking
+			let deadline = match self.try_lock_inner(self.deadline) {
+				Ok(_) => return StorageLockGuard::<'a, 'b, L> { lock: Some(self) },
+				Err(Some(other_locks_deadline)) => other_locks_deadline,
+				_ => L::deadline(), // use the default
+			};
+			L::current().snooze(&deadline);
+		}
+	}
+
+	/// Explicitly unlock the lock.
+	pub fn unlock(&mut self) {
+		self.value_ref.clear();
+	}
+}
+
+/// Extension trait for locks based on [`Timestamp`](::sp_core::offchain::Timestamp).
+///
+/// Allows explicity setting the timeout on construction
+/// instead of using the implicit default timeout of
+/// [`STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS`](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+pub trait TimeLock<'a>: Sized {
+	/// Provide an explicit deadline timestamp at which the locked state of the lock
+	/// becomes stale and may be dismissed by `fn try_lock(..)`, `fn spin_lock(..)` and others.
+	fn with_deadline<'k>(key: &'k [u8], lock_deadline: Timestamp) -> Self
+	where
+		'k: 'a;
+}
```

Comment:
```suggestion
pub trait TimeLock<'a>: Sized {
	/// Provide an explicit deadline timestamp at which the locked state of the lock
	/// becomes stale and may be dismissed by `fn try_lock(..)`, `fn spin_lock(..)` and others.
	fn with_deadline(key: &'a [u8], lock_deadline: Timestamp) -> Self;
}
```

Reply:
Thus `'k` is not even going anywhere, and is converted to `'a` because Self is `: TimeLock<'a. ..>`, is it / isn't it?

Otherwise can you write a test/example to show this is required?